src/sync/pull.ts

```diff
@@ -0,0 +1,403 @@
+  BeginTryPullRequest,
+  BeginTryPullResponse,
+  ChangedKeysMap,
+  HTTPRequestInfo,
+  MaybeEndTryPullRequest,
+  MaybeEndTryPullResponse,
+  ReplayMutation,
+} from '../repm-invoker.js';
+import {callJSRequest} from './js-request';
+import {SYNC_HEAD_NAME} from './sync-head-name.js';
+import * as patch from './patch.js';
+import * as prolly from '../prolly/mod.js';
+import type {LogContext} from '../rlog/logger.js';
+
+export const PULL_VERSION = 0;
+
+export type PullRequest = {
+  clientID: string;
+  cookie: JSONValue;
+  lastMutationID: number;
+  pullVersion: number;
+  // schema_version can optionally be used by the customer's app
+  // to indicate to the data layer what format of Client View the
+  // app understands.
+  schemaVersion: string;
+};
+
+export interface InternalPuller {
+  pull(
+    pullReq: PullRequest,
+    url: string,
+    auth: string,
+    requestID: string,
+  ): Promise<[PullResponse | undefined, HTTPRequestInfo]>;
+}
+
+export async function beginPull(
+  clientID: string,
+  beginPullReq: BeginTryPullRequest,
+  puller: InternalPuller,
+  requestID: string,
+  store: dag.Store,
+  lc: LogContext,
+): Promise<BeginTryPullResponse> {
+  const {pullURL, pullAuth, schemaVersion} = beginPullReq;
+
+  const baseSnapshot = await store.withRead(async dagRead => {
+    const mainHeadHash = await dagRead.getHead(db.DEFAULT_HEAD_NAME);
+    if (!mainHeadHash) {
+      throw new Error('Internal no main head found');
+    }
+    return await db.Commit.baseSnapshot(mainHeadHash, dagRead);
+  });
+
+  const [baseLastMutationID, baseCookie] =
+    db.Commit.snapshotMetaParts(baseSnapshot);
+
+  const pullReq = {
+    clientID,
+    cookie: baseCookie,
+    lastMutationID: baseSnapshot.mutationID(),
+    pullVersion: PULL_VERSION,
+    schemaVersion,
+  };
+  lc.debug?.('Starting pull...');
+  const pullStart = Date.now();
+  const [pullResp, httpRequestInfo] = await puller.pull(
+    pullReq,
+    pullURL,
+    pullAuth,
+    requestID,
+  );
+
+  lc.debug?.(
+    `...Pull ${pullResp ? 'complete' : 'failed'} in `,
+    Date.now() - pullStart,
+    'ms',
+  );
+
+  // If Puller did not get a pull response we still want to return the HTTP
+  // request info to the JS SDK.
+  if (!pullResp) {
+    return {
+      httpRequestInfo,
+      syncHead: '',
+      requestID,
+    };
+  }
+
+  // It is possible that another sync completed while we were pulling. Ensure
+  // that is not the case by re-checking the base snapshot.
+  return await store.withWrite(async dagWrite => {
+    const dagRead = dagWrite.read();
+    const mainHeadPostPull = await dagRead.getHead(db.DEFAULT_HEAD_NAME);
+
+    if (mainHeadPostPull === undefined) {
+      throw new Error('Main head disappeared');
+    }
+    const baseSnapshotPostPull = await db.Commit.baseSnapshot(
+      mainHeadPostPull,
+      dagRead,
+    );
+    if (baseSnapshot.chunk.hash !== baseSnapshotPostPull.chunk.hash) {
+      throw new Error('Overlapping syncs JSLogInfo');
```

Comment:
This line is never tested. I guess we should have a custom Error class for these?

But first we need to figure out how to test this...